rand_distr/src/triangular.rs

```diff
@@ -62,37 +62,165 @@ impl Distribution<f64> for Triangular {
         }
+        
+        let range = max - min;
+        let mu = (min + max + shape * mode) / (shape + 2.);
+        let v = if mu == mode {
+            shape * 0.5 + 1.
+        } else {
+            (mu - min) * (2. * mode - min - max)
+                / ((mode - mu) * (max - min))
+        };
+        let w = v * (max - mu) / (mu - min);
+        let beta = Beta::new(v, w).map_err(|_| PertError::RangeTooSmall)?;
+        Ok(Pert{ min, range, beta })
+    }
+}
+
+impl Distribution<f64> for Pert {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> f64 {
+        self.beta.sample(rng) * self.range + self.min
     }
 }
 
 #[cfg(test)]
 mod test {
-    use crate::Distribution;
-    use super::Triangular;
+    use std::f64;
+    use rand::{Rng, rngs::mock};
+    use super::*;
 
     #[test]
-    fn test_new() {
+    fn test_triangular() {
+        let mut half_rng = mock::StepRng::new(0x8000_0000_0000_0000, 0);
+        assert_eq!(half_rng.gen::<f64>(), 0.5);
+        for &(min, max, mode, median) in &[
+            (-1., 1., 0., 0.),
+            (1., 2., 1., 2. - 0.5f64.sqrt()),
+            (5., 25., 25., 5. + 200f64.sqrt()),
+            (1e-5, 1e5, 1e-3, 1e5 - 4999999949.5f64.sqrt()),
+            (0., 1., 0.9, 0.45f64.sqrt()),
+            (-4., -0.5, -2., -4.0 + 3.5f64.sqrt()),
+        ] {
+            println!("{} {} {} {}", min, max, mode, median);
+            let distr = Triangular::new(min, max, mode).unwrap();
+            // Test correct value at median:
+            assert_eq!(distr.sample(&mut half_rng), median);
+        }
+
         for &(min, max, mode) in &[
-            (-1., 1., 0.), (1., 2., 1.), (5., 25., 25.), (1e-5, 1e5, 1e-3),
-            (0., 1., 0.9), (-4., -0.5, -2.), (-13.039, 8.41, 1.17),
+            (-1., 1., 2.),
+            (-1., 1., -2.),
+            (2., 1., 1.),
         ] {
-            println!("{} {} {}", min, max, mode);
-            let _ = Triangular::new(min, max, mode).unwrap();
+            assert!(Triangular::new(min, max, mode).is_err());
         }
     }
+    
+    #[test]
+    fn test_triangular_vector() {
+        let rng = crate::test::rng(860);
+        let distr = Triangular::new(2., 10., 3.).unwrap();
+        let seq = distr.sample_iter(rng).take(5).collect::<Vec<f64>>();
+        println!("seq: {:?}", seq);
+        let expected = vec![4.941640229082449, 2.421447306833011,
+                4.5964271605527385, 2.789763631136542, 4.8014432067978445];
+        assert!(seq == expected);
+    }
 
     #[test]
-    fn test_sample() {
-        let norm = Triangular::new(0., 1., 0.5).unwrap();
-        let mut rng = crate::test::rng(1);
-        for _ in 0..1000 {
-            norm.sample(&mut rng);
+    fn test_pert() {
+        for &(min, max, mode) in &[
+            (-1., 1., 0.),
+            (1., 2., 1.),
+            (5., 25., 25.),
+        ] {
+            let _distr = Pert::new(min, max, mode).unwrap();
+            // TODO: test correctness
```

Comment:
Here is the correct test example I used:
```rust
#[test]
fn test_pert_sample() {
    let mut rng = crate::test::rng(860);
    let distr = Pert::new(45., 225., 90.).unwrap();
    let seq = distr.sample(&mut rng);
    let expected = 106.34517425256126;

    assert!(seq == expected);
}
```

Rough hand calculations as follows:
```
optimistic: 45
pessimistic: 225
most likely: 90

estimate average: (45+225+90)/3 = 120
estimate PERT: (45+225+4×90)/6 = 105
```

Reply:
I think though it's pure luck that the sample lies close to your estimate? This is a random number distribution driven by random input. Note that the triangular distribution is tested by forcing the RNG to output "half", however interpretation is dependent on distribution implementations (which for this one includes the Beta and Gamma distrs).

Reply:
Hmm, not really.. :thinking: 
![image](https://user-images.githubusercontent.com/8407846/56497897-bb1a1e80-64ff-11e9-9e34-af72e8ab81ba.png)

This results can be proven by Monte Carlo simulation:
![image](https://user-images.githubusercontent.com/8407846/56497872-a047aa00-64ff-11e9-839d-5e89a0b2858a.png)

Thus your `PERT` implementation is good enough for most cases, IMHO. :slightly_smiling_face:

Reply:
This still doesn't test correctness. The results from the included test vector are also quite close (μ = 4, var = 12/7), so I don't think including your example is useful.

Reply:
@dhardy I never asked for inclusion. :confused:  
Just decried my own experience checking out and testing the branch. :slightly_smiling_face: